server/portal/apps/projects/workspace_operations/shared_workspace_operations.py

```diff
@@ -0,0 +1,363 @@
+                              username: str,
+                              public_key: str,
+                              private_key: str,
+                              system_id) -> None:
+    """
+    Set an RSA key pair as the user"s auth credential on a Tapis system.
+    """
+    logger.info(f"Creating user credential for {username} on Tapis system {system_id}")
+    data = {"privateKey": private_key, "publicKey": public_key}
+    client.systems.createUserCredential(
+        systemId=system_id,
+        userName=username,
+        **data
+    )
+
+
+def register_public_key(username, publicKey, system_id) -> int:
+    """
+    Push a public key to the Key Service API.
+    """
+    url = "https://api.tacc.utexas.edu/keys/v2/" + username
+    headers = {"Authorization": "Bearer {}".format(settings.KEY_SERVICE_TOKEN)}
+    data = {"key_value": publicKey, "tags": [{"name": "system", "value": system_id}]}
+    response = requests.post(url, json=data, headers=headers)
+    response.raise_for_status
+    return response.status_code
+
+
+def set_workspace_permissions(client: Tapis, username: str, system_id: str, role: str):
+    """Apply read/write/execute permissions to a user on a system."""
+
+    system_pems = {
+         "reader": ["READ", "EXECUTE"],
+         "writer": ["READ", "EXECUTE"]
+    }
+
+    files_pems = {
+         "reader": "READ",
+         "writer": "MODIFY"
+    }
+
+    logger.info(f"Adding {username} permissions to Tapis system {system_id}")
+    client.systems.grantUserPerms(
+        systemId=system_id,
+        userName=username,
+        permissions=system_pems[role])
+
+    if role == "reader":
+        client.systems.revokeUserPerms(systemId=system_id,
+                                       userName=username,
+                                       permissions=["MODIFY"])
+        client.files.deletePermissions(systemId=system_id,
+                                       path="/",
+                                       username=username)
+
+    client.files.grantPermissions(
+        systemId=system_id,
+        path="/",
+        username=username,
+        permission=files_pems[role]
+    )
+
+
+def set_workspace_acls(client, system_id, path, username, operation, role):
+    jwt = client.access_token.access_token
+    url = f"{settings.TAPIS_TENANT_BASEURL}/v3/files/utils/linux/facl/{system_id}/{path}"
+    headers = {"x-tapis-token": jwt}
+
+    operation_map = {
+         "add": "ADD",
+         "remove": "REMOVE"
+    }
+
+    acl_string_map = {
+         "reader": f"d:u:{username}:rX,u:{username}:rX",
+         "writer": f"d:u:{username}:rwX,u:{username}:rwX",
+         "none": f"d:u:{username},u:{username}"
+    }
+
+    data = {"operation": operation_map[operation],
+            "aclString": acl_string_map[role],
+            "recursionMethod": "PHYSICAL"}
+
+    response = requests.post(url, json=data, headers=headers)
+    response.raise_for_status()
+    return response
```

Comment:
Probably, I wrote this util before we knew about the trick to get tapipy to fetch the API schema. Do you want to refactor this now or hold off until ACL stuff is baked into tapipy? I'm going to be honest that I don't see much actual value in factoring out `requests` calls. The main benefit you'd expect from an SDK is models/function signatures/implicit documentation, but tapipy is a total black box without the live docs.

Reply:
My main push for using tapipy is adoption of a common tapis request format. i.e. if Tapipy supports it, implicitly or otherwise, we should use it as a common pattern